### PR TITLE
Feature/CVSB-14939 - [FE] - As the DVSA, we want to be able to generate plates for technical records, so that we can supply plates with accurate information to the applicant

### PR DESCRIPTION
**Pull request by irina005**
## CVSB-14939 - generate plates for technical records
[FE] - As the DVSA, we want to be able to generate plates for technical records, so that we can supply plates with accurate information to the applicant

https://jira.dvsacloud.uk/browse/CVSB-14939

Bugs solved in here:

- https://jira.dvsacloud.uk/browse/CVSB-17099

## Checklist

- [x] Branch is rebased against the latest develop/common
- [x] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [x] PR title includes the JIRA ticket number
- [x] Squashed commits contain the JIRA ticket number
- [x] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { Component, OnInit, ChangeDetectionStrategy, Inject } from '@angular/core';
```
Comment (alleypa) [inline src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:1]: Please let catch up on this implementation, no sure why you splitted out the template

File: src/app/store/actions/VehicleTechRecordModel.actions.ts [hunk 2]
```diff
@@ -20,7 +20,9 @@ export enum EVehicleTechRecordModelActions {
   SetVehicleTechRecordModelOnCreate = '[VehicleTechRecordModel] SetVehicleTechRecordModelVinOnCreate',
   SetSelectedVehicleTechnicalRecord = '[MultipleTechRecordContainer] Set Selected Vehicle Technical Record',
   SetSelectedVehicleTechnicalRecordSucess = '[VehicleTechnicalRecordEffect] Set SelectedVehicleTechnicalRecordSucess',
-  SetViewState = '[Vehicle TechRecord Effect] Set ViewState'
+  SetViewState = '[Vehicle TechRecord Effect] Set ViewState',
+
+  SendMinistryPlates = '[Vehicle TechRecord Effect] Send Ministry Plates'
```
Comment (alleypa) [inline src/app/store/actions/VehicleTechRecordModel.actions.ts:25]: do you need this

File: src/app/technical-record/ministry-plates-template/ministry-plates-template.component.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    this.platesForm = new FormGroup({});
+    this.platesForm.addControl('plateReasonForIssue', new FormControl());
+
+    this.platesFormSubscription = this.platesForm.valueChanges.subscribe((result) => {
+      const plateReason = result.plateReasonForIssue.includes('original')
+        ? 'Original'
+        : 'Replacement';
+      this.newPlate = this.setPlateAttributes(plateReason);
+      this.plateHandler.emit(this.newPlate);
+    });
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates-template/ministry-plates-template.component.ts:43]: not sure you need `valueChanges` to grap the form data, You can grap all form update for creating the plate object when the user presses 'send ministry plate' and dialog is closed

File: src/app/technical-record/technical-record.component.html [hunk 4]
```diff
@@ -7,25 +7,52 @@
...
+          class="govuk-button govuk-button--secondary"
+          id="test-send-ministry-plates"
+          type="button"
+          data-module="govuk-button"
+          (click)="onSendMinistryPlates()"
+        >
+          Send ministry plates
+        </button>
+      </div>
+    </div>
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.html:55]: wrap in <ngcontainer> and put the ngIf on it.  leave the class on the div

File: src/app/technical-record/technical-record.component.ts [hunk 5]
```diff
@@ -23,6 +23,7 @@ import {
 } from '@app/models/vehicle-tech-record.model';
 import { TestResultModel } from '@app/models/test-result.model';
 import { VIEW_STATE } from '@app/app.enums';
+import { DialogMinistryPlatesComponent } from '@app/shared/dialog-ministry-plates/dialog-ministry-plates.component';
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:26]: move under this import `import { AdrReasonModalComponent } from '@app/shared/adr-reason-modal/adr-reason-modal.component';`

File: src/app/technical-record/technical-record.component.ts [hunk 6]
```diff
@@ -162,4 +164,22 @@ export class TechnicalRecordComponent implements OnChanges, OnInit {
...
     // }
   }
+
+  onSendMinistryPlates() {
+    const dialogRef = this.dialog.open(DialogMinistryPlatesComponent, {
+      width: '45vw',
+      data: { extra: this.activeRecord.applicantDetails }
+    });
+
+    dialogRef.afterClosed().subscribe((result) => {
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:174]: this is a plate object right? `result:plate`

File: src/app/utils/testing.utils.ts [hunk 7]
```diff
@@ -262,7 +262,8 @@ function mockPlates(args?: Partial<Plate>): Plate {
     plateSerialNumber: '123123',
     plateIssueDate: '2019-12-13',
     plateReasonForIssue: '5678',
-    plateIssuer: 'Issuer1'
+    plateIssuer: 'Issuer1',
+    toEmailAddress: 'test'
```
Comment (alleypa) [inline src/app/utils/testing.utils.ts:266]: ideally this update should be from your test,

File: src/app/technical-record/ministry-plates-template/ministry-plates-template.component.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  }
+  setPlateAttributes(plateReason: string): Plate {
+    return {
+      plateIssuer: this.loggedUser.getUser().msUser,
+      plateIssueDate: new Date().toISOString(),
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates-template/ministry-plates-template.component.ts:48]: you should have this format `2020-06-25` right?

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+export interface DialogData {
+  extra: Applicant;
+}
+
+@Component({
+  selector: 'vtm-dialog-ministry-plates',
+  templateUrl: './dialog-ministry-plates.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
```
Comment (alleypa) [inline src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:14]: `exportAs: plateComponent`

File: src/app/technical-record/ministry-plates-template/ministry-plates-template.component.html [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+          </label>
+        </div>
+      </div>
+    </fieldset>
+  </div>
+</form>
+<div class="govuk-inset-text">
+  A digital copy of the ministry plate will be email to the applicant on the tech record ({{
+    applicantDetails?.emailAddress
+  }})
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates-template/ministry-plates-template.component.html:25]: this bind should be part of the plate Form group

File: src/app/technical-record/technical-record.component.ts [hunk 11]
```diff
@@ -162,4 +164,22 @@ export class TechnicalRecordComponent implements OnChanges, OnInit {
...
+      if (result) {
+        if (!!this.activeVehicleTechRecord.techRecord[0].plates) {
+          this.activeVehicleTechRecord.techRecord[0].plates.push(result);
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:177]: grap a copy of `this.activeVehicleTechRecord` b4 mutating

File: src/app/technical-record/technical-record.component.ts [hunk 12]
```diff
@@ -162,4 +164,22 @@ export class TechnicalRecordComponent implements OnChanges, OnInit {
     });
     // }
   }
+
+  onSendMinistryPlates() {
+    const dialogRef = this.dialog.open(DialogMinistryPlatesComponent, {
+      width: '45vw',
+      data: { extra: this.activeRecord.applicantDetails }
+    });
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:172]: data: { applicantDetails: this.activeRecord.applicantDetails }

Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:172]: `data: { applicantDetails: this.activeRecord.applicantDetails }`

File: src/app/store/actions/VehicleTechRecordModel.actions.ts [hunk 13]
```diff
@@ -20,7 +20,7 @@ export enum EVehicleTechRecordModelActions {
   SetVehicleTechRecordModelOnCreate = '[VehicleTechRecordModel] SetVehicleTechRecordModelVinOnCreate',
   SetSelectedVehicleTechnicalRecord = '[MultipleTechRecordContainer] Set Selected Vehicle Technical Record',
   SetSelectedVehicleTechnicalRecordSucess = '[VehicleTechnicalRecordEffect] Set SelectedVehicleTechnicalRecordSucess',
-  SetViewState = '[Vehicle TechRecord Effect] Set ViewState'
+  SetViewState = '[Vehicle TechRecord Effect] Set ViewState',
```
Comment (alleypa) [inline src/app/store/actions/VehicleTechRecordModel.actions.ts:23]: undo line. one less file to review

File: src/app/technical-record/technical-record.component.ts [hunk 14]
```diff
@@ -166,4 +168,25 @@ export class TechnicalRecordComponent implements OnChanges, OnInit {
...
+    const dialogRef = this.dialog.open(DialogMinistryPlatesComponent, {
+      width: '45vw',
+      data: { applicantDetails: this.activeRecord.applicantDetails }
+    });
+
+    dialogRef.afterClosed().subscribe((plate) => {
+      const activeTechRec = JSON.parse(
+        JSON.stringify(this.activeVehicleTechRecord.techRecord[0])
+      );
+      if (plate) {
```
Comment (alleypa) [inline src/app/technical-record/technical-record.component.ts:182]: can we ever have a situation when the `plate` is empty or undefined

File: src/app/technical-record/ministry-plates/ministry-plates.component.ts [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  @Output() plateHandler = new EventEmitter<Plate>();
+  platesForm: FormGroup;
+  issueValues = Object.values(PLATE_ISSUE_VALUES);
+
+  constructor(public loggedUser: UserService) {}
+
+  ngOnInit() {
+    const plateIssuer = !!this.loggedUser.getUser()
+      ? this.loggedUser.getUser().msUser
+      : ({} as UserDetails);
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates/ministry-plates.component.ts:33]: `this.loggedUser.getUser().msUser` should be sufficient. The value should be sent for you to take this plate action.

File: src/app/technical-record/ministry-plates/ministry-plates.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+
+  beforeEach(async(() => {
+    getUser = jest.fn();
+    TestBed.configureTestingModule({
+      imports: [ReactiveFormsModule],
+      declarations: [MinistryPlatesComponent],
+      providers: [
+        {
+          provide: UserService,
+          useValue: { getUser }
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates/ministry-plates.component.spec.ts:22]: you need to provide a mock return value for the getUser

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+  ngOnInit() {}
+
+  close(): void {
+    this.dialogRef.close(false);
+  }
+
+  save(plateComponent): void {
+    const plate = !!plateComponent.platesForm
+      ? plateComponent.platesForm.getRawValue()
+      : ({} as Plate);
```
Comment (alleypa) [inline src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:30]: `const plate = plateComponent.platesForm.getRawValue() or plateComponent.platesForm.value`. Plate can never be empty right?

File: src/app/technical-record/ministry-plates/ministry-plates.component.ts [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+} from '@angular/core';
+import { FormControl, FormGroup } from '@angular/forms';
+
+import { PLATE_ISSUE_VALUES } from '@app/app.enums';
+import { UserService } from '@app/app-user.service';
+import { Applicant, Plate } from '@app/models/tech-record.model';
+import { UserDetails } from '@app/models/user-details';
+
+@Component({
+  selector: 'vtm-ministry-plates-template',
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates/ministry-plates.component.ts:17]: `vtm-ministry-plates` instead

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.html [hunk 19]
```diff
@@ -0,0 +1,33 @@
+<div class="govuk-grid-row">
```
Comment (alleypa) [inline src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.html:1]: we also need to run prettier to have consistent formatted code

File: src/app/technical-record/ministry-plates/ministry-plates.component.html [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+    <fieldset class="govuk-fieldset">
+      <div class="govuk-radios">
+        <div *ngFor="let issue of issueValues; let i = index" class="govuk-radios__item">
+          <input
+            class="govuk-radios__input"
+            id="test-issue-{{ i }}"
+            type="radio"
+            formControlName="plateReasonForIssue"
+            value="{{ issue.includes('original') ? 'Original' : 'Replacement' }}"
+          />
```
Comment (alleypa) [inline src/app/technical-record/ministry-plates/ministry-plates.component.html:13]: Let discuss this. but ideally you should covert your radio choices to options so you wont need to do this `value="{{ issue.includes('original') ? 'Original' : 'Replacement' }}"`

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts (hunk 1 above)
Comment (irina005) [reply src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:1]: Please see the exportAs in the ministry-plates.component.ts

Comment (irina005) [reply src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:1]: Closing as I added exportAs in ministry-plates.component.ts
image

File: src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts (hunk 9 above)
Comment (irina005) [reply src/app/shared/dialog-ministry-plates/dialog-ministry-plates.component.ts:14]: Closing as I added exportAs in ministry-plates.component.ts
image

File: src/app/technical-record/technical-record.component.html [hunk 21]
```diff
@@ -7,25 +7,53 @@
...
+        <ng-container>
+          <button
+            id="test-save-btn"
+            class="govuk-button"
+            type="submit"
+            *ngIf="editState || createState"
+          >
+            Save technical record
+          </button>
+          <a id="cancel2" (click)="cancelTechRecord()" *ngIf="editState">
```
Comment (JonLazariniDD) [inline src/app/technical-record/technical-record.component.html:32]: can we prepend with it with test- please?

**Comment by JonLazariniDD** [comment]
Looks good, just the id to be checked with the testers I think and run the regression pack.

**Pull request merged**